### PR TITLE
feat(adapters): new EdgeDB Adapter

### LINKED ISSUE #4730: Add edgedb adapter
### Description 📓

Edgedb is an awesome graph database. It reached 1.0 not so long ago and it has a similar developer experience as prisma, with a type-safe query builder. It would be awesome to include an adapter for it. Here's a [link](https://www.edgedb.com/docs/clients/01_js/querybuilder) if you're not familiar with it.

### How to reproduce ☕️

```ts
export default function EdgedbAdapter(client, options = {}) {
  return {
    async createUser(user) {
      return client.querySingle(`insert User {
        name := <str>$name
        email := <str>$email
      }`, user)
    },
    // ...rest of the methods
  }
}
```

### Contributing 🙌🏽

Yes, I am willing to help implement this feature in a PR

### PR DESCRIPTION
**Pull request by brunocrosier**
## ☕️ Reasoning

Adds an adapter for [EdgeDB](https://edgedb.com).

## 🧢 Checklist

- [X] Documentation
- [X] Tests
- [X] Ready to be merged

## 🎫 Affected issues

Fixes: https://github.com/nextauthjs/next-auth/issues/4730

## 📌 Resources

- [Security guidelines](../Security.md)
- [Contributing guidelines](../CONTRIBUTING.md)
- [Code of conduct](../CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

### CODE REVIEW

**Comment by brunocrosier** [comment]
Everything inside of `packages/adapter-edgedb/dbschema/edgeql-js` is generated by `npx @edgedb/generate edgeql-js` 

Tests are all passing:

image

**Comment by 1st1** [comment]
Impressive!

I don't think we should make the querybuilder artifacts part of this PR. That's generally a bad idea because things change from version to version of edgedb-js. Users instead should run the generator themselves.

**Comment by brunocrosier** [comment]
> Impressive!
> 
> I don't think we should make the querybuilder artifacts part of this PR. That's generally a bad idea because things change from version to version of edgedb-js. Users instead should run the generator themselves.

thanks! yeah, i was curious about whether to include these files or not.. 

the problem is that it we don't check these files in, anyone who wants to develop the adapter in the future would need to run the generate script before building the package. 

maybe a solution could be if i add a `postinstall` script which runs `npx @edgedb/generate edgeql-js` ?

File: apps/dev/dbschema/futures.esdl
Comment (colinhacks) [inline apps/dev/dbschema/futures.esdl:4]: I have a slight preference for keeping all the schema information in one file but your call 👍

File: docs/docs/adapters/edgedb.md
Comment (colinhacks) [inline docs/docs/adapters/edgedb.md:15]: It might be mroe future-proof to link out to our quickstart guide. Something like 
"Follow the EdgeDB quickstart to install the EdgeDB CLI and initailize a project". That way you don't need to worry about keeping all this information up-to-date if/when workflows change.

File: docs/docs/adapters/edgedb.md
Comment (colinhacks) [inline docs/docs/adapters/edgedb.md:69]:
```suggestion
Replace the contents of the auto-generated file in `dbschema/default.esdl` with the following:
```

File: docs/docs/adapters/edgedb.md
Comment (colinhacks) [inline docs/docs/adapters/edgedb.md:217]: nice

File: docs/docs/adapters/edgedb.md
Comment (colinhacks) [inline docs/docs/adapters/edgedb.md:214]: Might be worth noting that the exact mechanism for setting this variable will depend on the hosting platform...e.g. with Vercel you set this through the dashboard.

**Comment by colinhacks** [comment]
> the problem is that it we don't check these files in, anyone who wants to develop the adapter in the future would need to run the generate script before building the package

Hate to say it, but it probably makes more sense for adapters like this to just use raw EdgeQL instead of the query builder. It solves all the issues w.r.t. generated files and won't be affected by future API/workflow changes. For stuff like this I think it's the wisest call.

File: apps/dev/dbschema/default.esdl
Comment (jaclarke) [inline apps/dev/dbschema/default.esdl:17]: The `userId` field here and on `Session` seem redundant, since there's already a link to `User`. If `userId` is needed in a query, it should be a computed property instead.

File: packages/adapter-edgedb/src/index.ts
Comment (jaclarke) [inline packages/adapter-edgedb/src/index.ts:61]: I don't think `assert_single` is needed, `filter_single` already implies that.

File: packages/adapter-edgedb/src/index.ts
Comment (jaclarke) [inline packages/adapter-edgedb/src/index.ts:93]: You can wrap the `update` query above in this `select` to update and select in a single query.

File: packages/adapter-edgedb/src/index.ts
Comment (jaclarke) [inline packages/adapter-edgedb/src/index.ts:210]: I think this should update the `user` link (and the `userId` should be computed)

**Comment by brunocrosier** [comment]
thanks for the suggestions - i've refactored the PR to use pure EdgeQL..

tests are passing and i've tested it with the Email and GitHub providers, all seemed to work OK 🤞🏼

**Comment by brunocrosier** [comment]
FYI i've published the package to https://www.npmjs.com/package/@bruno_crosier/edgedb-adapter so you can test it out locally and follow the instructions in the documentation (https://next-auth-git-fork-brunocrosier-feat-edgedb-adapter-nextauthjs.vercel.app/adapters/edgedb)

Of course, you'll need to just make sure you install `@bruno_crosier/edgedb-adapter` instead of `next-auth/edgedb-adapter` for now!

**Comment by Lucasvo1** [comment]
How is the performance versus using the Prisma adapter? 

Would Edgedb enable us to use middleware with database persisted sessions?

**Comment by brunocrosier** [comment]
> How is the performance versus using the Prisma adapter?

EdgeDB performance vs Prisma and other ORMs: https://edgedb.github.io/imdbench/js.html

> Would Edgedb enable us to use middleware with database persisted sessions?

Do you mean a Next.js middleware?

**Comment by Lucasvo1** [comment]
Yes, as far as I understand currently the Nextjs middleware route protection with NextAuth only works with JWTs, not with any adapters.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

File: packages/adapter-edgedb/src/index.ts
Comment (brunocrosier) [reply packages/adapter-edgedb/src/index.ts:19]: no! good catch!

File: packages/adapter-edgedb/src/index.ts
Comment (jaclarke) [inline packages/adapter-edgedb/src/index.ts:73]: Wrapping the query in `assert_single(...)` instead of using `limit 1` is the recommended best practice here, because `limit 1` will just silently discard results, instead of failing if the query does somehow return more than one `User`. Though since there's an exclusive constraint on `(.provider, .providerAccountId)`, querying the `Account` first then selecting the `user` link on it should let EdgeDB infer it's single without the need for `assert_single`. eg.
```
with account := (
  select Account
  filter .providerAccountId = <str>$providerAccountId
     and .provider = <str>$provider
)
select account.user {
  id,
  email,
  image,
  name,
  emailVerified
}
```

File: docs/docs/adapters/edgedb.md
Comment (jaclarke) [inline docs/docs/adapters/edgedb.md:12]: Now you're using raw edgeql queries, I don't think `@edgedb/generate` is needed any more right?

File: docs/docs/adapters/edgedb.md
Comment (jaclarke) [inline docs/docs/adapters/edgedb.md:162]: Same as above.

File: docs/docs/adapters/edgedb.md
Comment (brunocrosier) [reply docs/docs/adapters/edgedb.md:12]: We're using raw edgeql queries for the adapter itself, but I think we should still recommend developers to use `@edgedb/generate` in their code - right? As per: https://www.edgedb.com/docs/guides/tutorials/nextjs#generating-the-query-builder

In other words.. this documentation is recommending developers to install `@edgedb/generate` to their own code (along with the newly-created `@next-auth/edgedb-adapter`)

You can see how I've added this to the "demo" app here: https://github.com/nextauthjs/next-auth/pull/5781/files/97c9e801ea6d925d77cbb1846675407aeb26ceb7#diff-9ba7e4d9762adead1effbf9a9a1a94b85191064e2f9786da86b7918a7629e27d

File: docs/docs/adapters/edgedb.md
Comment (brunocrosier) [reply docs/docs/adapters/edgedb.md:162]: Same as above

**Comment by jaclarke** [comment]
@brunocrosier Other than the few minor comments I've posted, it all looks good to me.

File: packages/adapter-edgedb/src/index.ts
Comment (brunocrosier) [reply packages/adapter-edgedb/src/index.ts]: good point - thanks!

**Comment by brunocrosier** [comment]
Think this is good to go - thanks to reviews from @jaclarke and @colinhacks ! 

Could we get eyes on this from maintainers please @ThangHuuVu / @balazsorban44 ? 🙏🏼

**Review by ThangHuuVu — CHANGES_REQUESTED** [review]
Thanks & sorry for the delayed response
I added some review comments about moving stuffs around and a logic in `useVerificationToken`, could you check and resolve the conflict? 🙌

File: packages/adapter-edgedb/tsconfig.eslint.json
Comment (ThangHuuVu) [inline packages/adapter-edgedb/tsconfig.eslint.json:3]: This file isn't needed anymore, could you remove it? 🙌 We only use one `tsconfig.json` file for each package now

File: packages/adapter-edgedb/dbschema/default.esdl
Comment (ThangHuuVu) [inline packages/adapter-edgedb/dbschema/default.esdl:1]: We don't need this file in the adapter package, right?
Everything under `/dbschema` can be removed IMO

File: packages/adapter-edgedb/package.json
Comment (ThangHuuVu) [inline packages/adapter-edgedb/package.json:5]:
```suggestion
  "homepage": "https://authjs.dev",
```

File: packages/adapter-edgedb/package.json
Comment (ThangHuuVu) [inline packages/adapter-edgedb/package.json:45]: I don't think we need direct dependency - we already got `edgedb` defined at `peerDependencies`

**Review by ThangHuuVu — DISMISSED** [review]

File: apps/dev/nextjs/.vscode/settings.json
Comment (omavi) [inline apps/dev/nextjs/.vscode/settings.json:2]: Should this change be included in the PR? Looks like a Windows specific change.

File: apps/dev/nextjs/dbschema/migrations/00001.edgeql
Comment (omavi) [inline apps/dev/nextjs/dbschema/migrations/00001.edgeql:1]: Is the `migrations` folder necessary?

File: apps/dev/nextjs/pages/api/auth/[...nextauth].ts
Comment (omavi) [inline apps/dev/nextjs/pages/api/auth/[...nextauth].ts:91]: Why is Auth0 commented out here?

Comment (brunocrosier) [reply apps/dev/nextjs/pages/api/auth/[...nextauth].ts]: thanks - i commented this out by accident when testing (was trying to run the nextjs demo app locally and didn't want to set up Auth0 env var), resolved now!

File: apps/dev/nextjs/.vscode/settings.json
Comment (brunocrosier) [reply apps/dev/nextjs/.vscode/settings.json]: weird, didn't change this manually - maybe happened automatically when i ran `pnpm i` - changed back now!

**Comment by brielov** [comment]
Hi there! Any news on this? It would be awesome to have it merged soon. Anyway, thank you so much for taking the time to contribute @brunocrosier !

File: packages/adapter-test/index.ts
Comment (ndom91) [inline packages/adapter-test/index.ts:303]: Why are you adding all this logic to skip the `deleteUser` test? It does seem to be implemented here, right?

File: packages/next-auth/package.json
Comment (ndom91) [inline packages/next-auth/package.json:42]: Can we use [rimraf](https://www.npmjs.com/package/rimraf) instead maybe?

File: packages/adapter-test/index.ts
Comment (omavi) [reply packages/adapter-test/index.ts:303]: @ndom91 I'm a bit confused. The history shows that you added the logic you're asking to be removed. Can you explain?

**Comment by brielov** [comment]
Any reason this is being delayed for so long? Would be really nice to have edgedb as part of the ecosystem.

**Comment by brunocrosier** [comment]
> Any reason this is being delayed for so long? Would be really nice to have edgedb as part of the ecosystem.

I don't know why it's taking so long to be merged, I created the PR months ago. Maybe @ThangHuuVu can help?

**Comment by brielov** [comment]
Come on guys, this PR has been sitting here waiting to be merged for more than 6 months now. What can we do to get this going? 

Also a suggestion for @brunocrosier it seems that each query fetching the user has every field hard coded. I've come across this issue when you add extra fields to the schema and expect them to come back inside the `session` callback for example. I might be mistaken but I think the upcoming edgedb 3.0 version will support a wildcard-style eg: `select User {*}` for selecting everything, which I think it would be a nice thing to have. Anyway, I can only speak for myself but I think this should be in core for others to use.

**Comment by notramo** [comment]
@balazsorban44, what is going on with this?

**Review by ThangHuuVu — APPROVED** [review]

**Comment by ThangHuuVu** [comment]
merged now, thanks everyone for helping & sorry for the delay!

**Comment by ThangHuuVu** [comment]
PS the first version is published [here](https://github.com/nextauthjs/next-auth/releases/tag/%40auth%2Fedgedb-adapter%400.2.0), please go try it and give us feedback, thanks!

**Pull request merged**
